californium-core/src/main/java/org/eclipse/californium/core/network/stack/TCPBlockwiseLayer.java

```diff
@@ -0,0 +1,394 @@
+
+		BlockOption block1 = request.getOptions().getBlock1();
+
+		if ((block1 != null) && (block1.getSzx() == BERT_SZX)) {
+			KeyUri key = getKey(exchange, request);
+			Block1BlockwiseStatus status = getInboundBlock1Status(key, exchange, request);
+
+			if (!status.addBlock(request.getPayload())) {
+				sendBlock1ErrorResponse(key, exchange, request, ResponseCode.REQUEST_ENTITY_TOO_LARGE,
+						"body exceeded expected size " + status.getBufferSize());
+			}
+
+			status.setCurrentNum(status.getCurrentNum() + (request.getPayloadSize() / BERT_INT_BLOCK_SIZE));
+			if (block1.isM()) {
+
+				// Message has more blocks.
+
+				LOGGER.log(Level.FINE, "acknowledging incoming block1 [num={0}], expecting more blocks to come",
+						block1.getNum());
+
+				Response piggybacked = Response.createResponse(request, ResponseCode.CONTINUE);
+				piggybacked.getOptions().setBlock1(block1.getSzx(), true, block1.getNum());
+				piggybacked.setLast(false);
+
+				exchange.setCurrentResponse(piggybacked);
+				lower().sendResponse(exchange, piggybacked);
+
+			} else {
+
+				LOGGER.log(Level.FINE, "peer has sent last block1 [num={0}], delivering request to application layer",
+						block1.getNum());
+
+				// Remember block to acknowledge. TODO: We might make
+				// this a
+				// boolean flag in status.
+				exchange.setBlock1ToAck(block1);
+
+				// Assemble and deliver
+				Request assembled = new Request(request.getCode());
+				assembled.setSenderIdentity(request.getSenderIdentity());
+				status.assembleMessage(assembled);
+
+				assembled.setMID(request.getMID());
+				assembled.setToken(request.getToken());
+
+				assembled.getOptions().setBlock2(request.getOptions().getBlock2());
+
+				clearBlock1Status(key);
+
+				exchange.setRequest(assembled);
+				upper().receiveRequest(exchange, assembled);
+			}
+		} else {
+			super.handleInboundBlockwiseUpload(exchange, request);
+		}
+
+	}
+
+	/**
+	 * Handles the request for next block from the peer. Handling is done here
+	 * only for SZX = 7 all other handling is done in the BlockwikseLayer.
+	 */
+	protected void handleInboundRequestForNextBlock(final Exchange exchange, final Request request, final KeyUri key,
+			final Block2BlockwiseStatus status) {
+
+		BlockOption block2 = request.getOptions().getBlock2();
+		if ((block2 != null) && (block2.getSzx() == BERT_SZX)) {
+
+			synchronized (status) {
+
+				ByteBuffer responseBlockBuilder = ByteBuffer.allocate(bertStepSize * BERT_INT_BLOCK_SIZE);
+				int blockNum = block2.getNum();
+				int currentNum = blockNum;
+				boolean hasNextBlock = true;
+				Response block = null;
+
+				for (int i = 0; ((i < bertStepSize) && (hasNextBlock)); i++) {
+					block = status.getNextResponseBlock(block2);
+					hasNextBlock = block.getOptions().getBlock2().isM();
+					currentNum = currentNum + 1;
+					block2 = new BlockOption(BERT_SZX, hasNextBlock, currentNum);
+					responseBlockBuilder.put(block.getPayload());
+				}
+
```

Comment:
The idea is to have minimal changes in BlockwiseLayer code as UDP has greater priority than TCP at the moment.